### DISCUSSION TITLE
Make `startTrip` independent by combining it with `startTracking` and `createTrip` methods

### DISCUSSION THREAD

**Discussion by jothipriyadharshanr** [roam-ai/roam-ios#27] [Ideas]
### Discussed in https://github.com/roam-ai/roam-ios/discussions/27

<div type='discussions-op-text'>

<sup>Originally posted by **jothipriyadharshanr** August 10, 2021</sup>
**Currently**
- The `startTrip()` method has a mandatory input parameter ie. `trip_id` and return success or error callbacks when trips starts or error.
- The restriction for the developer is that they need a valid `trip_id` which has to be either created by `createTrip()` method or Create Trip API assigned to the user initially before `startTrip()`
- And if the developer wants to `startTracking()` with desired tracking method along with the `startTrip()` is not possible. 

**Proposed**
- We need the `startTrip()` method to have optional values below:
- [ ] trackingMode
- [ ] description
- [ ] metaData
- [ ] localTrip
- [ ] origin(s) & destination(s) 
- [ ] tripId
- We also need the `stopTrip()` method to have optional values below:
- [ ] tripId
- [ ] trackingMode
- [ ] forceStop

**Example**
```swift
Roam.startTrip(tripId: String,  metaData: Dictionary, localTrip: Bool, origin: Dictionary, destination: Dictionary, description: String, trackingMethod: RoamTrackingMode, options: RoamTrackingCustomMethods ) { RoamStartTrip, RoamError in
            RoamStartTrip.tripId,
            RoamStartTrip.status,
            RoamStartTrip.createdAt,
            RoamStartTrip.updatedAt,
            RoamStartTrip.userId,
            RoamStartTrip.tripTrackingUrl,
            RoamStartTrip.origin,
            RoamStartTrip.destination,
            RoamStartTrip.isStarted,
            RoamStartTrip.isDeleted,
            RoamStartTrip.isEnded,
            RoamStartTrip.metaData,
            RoamStartTrip.isLocalTrip,
            RoamStartTrip.description
        }
```

```swift
Roam.endTrip(tripId: String, trackingMethod: RoamTrackingMode, options: RoamTrackingCustomMethods, forceStopTracking: Bool) { RoamEndTrip, RoamError in
            RoamEndTrip.status
            RoamEndTrip.tripId
        }
```
</div>

**Comment by nikhil290** [comment]
@jothipriyadharshanr 
 localTrip is optional ?Then if user does not pass any value what will be the default value to if trip is online or offline ?
cc : @ravikanth-tummala @BirjuKr

**Reply by jothipriyadharshanr** [reply]
@nikhil290 I will update the optional and mandatory information in the main topic.

**Reply by jothipriyadharshanr** [reply]
- We need the `startTrip()` method to have optional values below:
- [ ] trackingMode *optional and defaults to 'Balanced' tracking mode if none is passed and tracking was not already started  
- [ ] description * optional
- [ ] metaData * optional
- [ ] localTrip * optional and will be ignored if tripId is passed ie. trip already created
- [ ] origin(s) & destination(s)  * optional and will be ignored if tripId is passed ie. trip already created
- [ ] tripId * optional
- We also need the `stopTrip()` method to have optional values below:
- [ ] tripId * mandatory
- [ ] trackingMode * optional and will be ignored if forStopTracking is passed
- [ ] forceStopTracking * optional

**Reply by jothipriyadharshanr** [reply]
@nikhil290 Let me know if this helps.

**Reply by nikhil290** [reply]
Hi @jothipriyadharshanr I am not clear on "localTrip " 
consider id localTrip and tripid both not passed in that case how will we know if it is online trip or offline trip ?

**Reply by jothipriyadharshanr** [reply]
I did mention that `localTrip` value is optional and will be ignored if `tripId` is also passed ie. trip already created
